File: frontend/src/lib/schemas/framework12.ts
import { z } from 'zod';

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

export const framework12StatusSchema = z.enum([
  'SIZED',
  'WATCHLIST',
  'BLOCKED',
  'UNKNOWN',
]);
export type Framework12Status = z.infer<typeof framework12StatusSchema>;

// ---------------------------------------------------------------------------
// Decision matrix row
// ---------------------------------------------------------------------------

export const decisionMatrixRowSchema = z.object({
  priority_code: z.string(),
  priority_label: z.string(),
  track: z.string().nullable(),
  earnings_max_days: z.number().nullable(),
  earnings_min_days: z.number().nullable(),
  override_required: z.boolean(),
  underweight_required: z.boolean(),
  strong_flow_required: z.boolean(),
  size_min_pct: z.number(),
  size_max_pct: z.number(),
  timing_rule: z.string(),
  is_watchlist_only: z.boolean(),
  priority_order: z.number(),
  active: z.boolean(),
});
export type DecisionMatrixRow = z.infer<typeof decisionMatrixRowSchema>;

// ---------------------------------------------------------------------------
// Top-level result
// ---------------------------------------------------------------------------

export const framework12ResultSchema = z.object({
  ticker: z.string(),
  status: framework12StatusSchema,
  matched_row: decisionMatrixRowSchema.nullable(),
  current_nav_usd: z.number().nullable(),
  size_min_usd: z.number().nullable(),
  size_max_usd: z.number().nullable(),
  timing_rule: z.string().nullable(),
  blocked_reason: z.string().nullable(),
  evaluated_at: z.string(),
});
export type Framework12Result = z.infer<typeof framework12ResultSchema>;

export const framework12SizingSchema = z.object({
  ticker: z.string(),
  status: framework12StatusSchema,
  priority_code: z.string().nullable(),
  size_min_usd: z.number().nullable(),
  size_max_usd: z.number().nullable(),
  timing_rule: z.string().nullable(),
  current_nav_usd: z.number().nullable(),
  blocked_reason: z.string().nullable(),
});
export type Framework12Sizing = z.infer<typeof framework12SizingSchema>;

File: backend/src/atlas/schemas/framework12.py
"""Pydantic schemas for Framework 12 — Decision Matrix sizing.

Framework 12 only runs after Section 16 returns gate=PASS.  It chooses the
highest-priority matching row from the framework12_decision_matrix table and
returns the recommended USD position size based on live NAV from F30.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel

Framework12Status = Literal["SIZED", "WATCHLIST", "BLOCKED", "UNKNOWN"]


class DecisionMatrixRow(BaseModel):
    """One priority row from the decision matrix."""

    priority_code: str
    priority_label: str
    track: str | None = None
    earnings_max_days: int | None = None
    earnings_min_days: int | None = None
    override_required: bool
    underweight_required: bool
    strong_flow_required: bool
    size_min_pct: float
    size_max_pct: float
    timing_rule: str
    is_watchlist_only: bool
    priority_order: int
    active: bool


class Framework12Result(BaseModel):
    """Sizing verdict for a ticker that has cleared Section 16."""

    ticker: str
    status: Framework12Status

    # The decision-matrix row that drove the sizing (None if BLOCKED).
    matched_row: DecisionMatrixRow | None = None

    # Live NAV from Framework 30 at the time of evaluation.
    current_nav_usd: float | None = None

    # Recommended USD position size range — derived from row pct + NAV.
    size_min_usd: float | None = None
    size_max_usd: float | None = None

    # Operator-facing timing instruction copied from the matched row.
    timing_rule: str | None = None

    # When Section 16 did not pass, this carries the reason.
    blocked_reason: str | None = None

    evaluated_at: datetime
